packages/bundler/src/modules/DepositManager.ts

```diff
@@ -0,0 +1,51 @@
+import { BigNumber } from 'ethers'
+import { getUserOpMaxCost, IEntryPoint, requireCond, UserOperation, ValidationErrors } from '@account-abstraction/utils'
+import { MempoolManager } from './MempoolManager'
+
+/**
+ * manage paymaster deposits, to make sure a paymaster has enough gas for all its pending transaction in the mempool
+ * [EREP-010]
+ */
+export class DepositManager {
+  deposits: { [addr: string]: BigNumber } = {}
+
+  constructor (readonly entryPoint: IEntryPoint, readonly mempool: MempoolManager) {
+  }
+
+  async checkPaymasterDeposit (userOp: UserOperation): Promise<void> {
+    const paymaster = userOp.paymaster
+    if (paymaster == null) {
+      return
+    }
+    let deposit = await this.getCachedDeposit(paymaster)
+    deposit = deposit.sub(getUserOpMaxCost(userOp))
+
+    for (const entry of this.mempool.getMempool()) {
+      if (entry.userOp.paymaster === paymaster) {
+        deposit =
+          deposit.sub(BigNumber.from(getUserOpMaxCost(userOp)))
+      }
+    }
+
+    // [EREP-010] paymaster is required to have balance for all its pending transactions.
+    // on-chain AA31 checks the deposit for the current userop.
+    // but submitting all these UserOps it will eventually abort on this error,
+    // so it's fine to return the same code.
+    requireCond(deposit.gte(0), 'paymaster deposit too low for all mempool UserOps', ValidationErrors.BannedOrThrottledPaymaster)
+  }
+
+  /**
+   * clear deposits after some known change on-chain
+   */
+  clearCache (): void {
```

Comment:
I think it is a bad idea to have the cache stored in `DepositManager` but trigger the clearing of the cache from the `EventsManager`. I suggest you keep the entire logic of the cache inside this class, i.e. let the `getCachedDeposit()` function do the `clearCache` once every X blocks or whatever.

Reply:
the DepositManager manages the deposits. the simplest impl is to read from the blockchain everytime.
Instead, I wanted some optimization, and cache it until there is a network change - which is always signaled by a message.